boe_test.py

```diff
@@ -0,0 +1,238 @@
+        )
+
+    def evaluate(self, sols):
+        if not sols.shape[1] == self.solution_dim:
+            raise ValueError(
+                f"Expects sols to have shape (,{self.solution_dim}), actually "
+                f"gets shape {sols.shape}")
+
+        displacement = sols - self.shift
+        raw_obj = np.sum(np.square(displacement), axis=1)
+        objs = ((raw_obj - self._worst_obj) /
+                (self._best_obj - self._worst_obj) * 100)
+
+        clipped = sols.copy()
+        clip_indices = np.where(np.logical_or(clipped > 5.12, clipped < -5.12))
+        clipped[clip_indices] = 5.12 / clipped[clip_indices]
+        measures = np.concatenate(
+            (
+                np.sum(
+                    clipped[:, :self.solution_dim // 2], axis=1, keepdims=True),
+                np.sum(
+                    clipped[:, self.solution_dim // 2:], axis=1, keepdims=True),
+            ),
+            axis=1,
+        )
+
+        return objs, measures
+
+
+def save_heatmap(archive, heatmap_path):
+    plt.figure(figsize=(8, 6))
+    grid_archive_heatmap(archive, vmin=0, vmax=100, cmap="viridis")
+    plt.tight_layout()
+    plt.savefig(heatmap_path)
+    plt.close(plt.gcf())
+
+
+# logdir for saving experiment data
+trial_outdir = os.path.join(
+    os.path.curdir,
+    "test_logs",
+)
+trial_path = Path(trial_outdir)
+if not trial_path.is_dir():
+    trial_path.mkdir()
+
+
+# experiment params
+@dataclass
+class Params:
+    init_nsamples: int
+    total_itrs: int
+    search_nrestarts: int
+    entropy_ejie: bool
+    upscale_schedule: List[List]
+    batch_size: int
+    num_emitters: int
+    seed: int
+    logdir: str
+    log_evry: int
+
+
+params = Params(
+    init_nsamples=20,
+    total_itrs=1000,
+    search_nrestarts=5,
+    entropy_ejie=False,
+    upscale_schedule=[[5, 5], [10, 10], [25, 25]],
+    batch_size=1,
+    num_emitters=1,
+    seed=42,
+    logdir=trial_outdir,
+    log_evry=20,
+)
+
+wandb_logger = wandb.init(
+    project="BOP-Elites(sphere)",
+    config=asdict(params),
+    name="BOP-Elites",
+    id=str(uuid.uuid5(uuid.NAMESPACE_DNS, str(asdict(params)))),
+    tags=["BOP-Elites", "Sphere"],
+    resume="allow",  # Allow resuming from same run ID
+)
+
+# Sphere domain evaluator
+sphere_domain = Sphere(solution_dim=4, shift=2)
+
+reload_checkpoint = None
+if reload_checkpoint is None:
+    start_itr = 0
+    # The main grid archive that interacts with BOP-Elites. We start at the
+    # lowest resolution from ``upscale_schedule``.
+    archive = GridArchive(
+        solution_dim=sphere_domain.solution_dim,
+        dims=params.upscale_schedule[0],
+        ranges=[[-5.12, 5.12]] * 2,
+        seed=params.seed,
+    )
+    # The passive archive that does NOT interact with BOP-Elites other than
+    # storing all evaluated data seen so far. We do not scale the passive
+    # archive, and it always stays at the final resolution.
+    passive_archive = GridArchive(
```

Comment:
Is it possible for this to be passed as the `result_archive` to the scheduler? That would be more consistent with how CMA-MAE currently operates.